RFC-001-libp2p.adoc

```diff
@@ -0,0 +1,361 @@
+This section first briefly explains, how such multiplexing can be achieved if all you have are plain network sockets.
+As a second step, we will quickly cover, how the `substream` mechanism of libp2p affects this and thereby simplifies our application protocol.
+
+==== Multiplexing with plain sockets
+
+With traditional network sockets, there is always a listener and a dialer.
+The listening party has to have a port open the dialling party can connect to.
+If one party is behind a NAT, the only way to establish a network connection is for this party to dial to the other one.
+Establishing such a connection is expensive.
+We have to wait for the TCP handshake to complete and if there is transport encryption in place, keys might need to be exchanged before the first actual payload can be sent to the other party.
+Due to this cost and network limitations like NAT, it is good to reuse a connection between two parties once we have established one.
+This is where multiplexing comes into play.
+If both parties want to be able to send and receive messages at any point in time, those messages need to carry metadata so that the receiving end can correctly identify and associate messages.
+
+If all you have are plain network sockets, such a multiplexing strategy needs to be backed into your application protocol, which is a non-trivial increase in complexity.
+In fact, a former version of the COMIT messaging protocol (BAM!) had support for such functionality because we were not yet building on top of libp2p back then.
+
+==== Multiplexing with libp2p substreams
+
+Libp2p has the concept of substreams.
+A substream is a virtual socket for a specific protocol between two nodes.
+Opening a new substream is cheap compared to an actual network connection.
+It is also independent of the underlying physical network topology, i.e. it does not matter which party opened the actual physical network connection:
+Every party can open a new substream at any point in time.
+
+=== Design
+
+This paragraph describes the over-arching design of the COMIT messaging protocol.
+It builds heavily upon the https://github.com/libp2p/specs/tree/master/connections#definitions[terminology] that is used within the libp2p spec.
+
+To send a new message to another node, a COMIT node MUST open a new substream.
+Within a single substream, there is always just one _active_ party.
+Only the _active_ party is allowed to send a message.
+The dialer of the substream is initially the _active_ party.
+Hence, the node opening the substream is the first one to send a message.
+
+==== Message format
+
+All messages within the COMIT protocol are JSON objects.
+They MUST be serialized to a single line and are delimited by the newline character `\n`.
+The messages MUST use UTF-8 encoding.
+
+TODO: include pretty printing note
```

Comment:
❓When is it intended to resolve this TODO?

Reply:
Within this PR :sweat_smile: